src/goto-symex/goto_symex_state.cpp

```diff
@@ -796,28 +796,18 @@ ssa_exprt goto_symex_statet::add_object(
 {
   framet &frame = call_stack().top();
 
-  ssa_exprt ssa = rename_ssa<L0>(ssa_exprt{expr}, ns).get();
-  const irep_idt l0_name = ssa.get_identifier();
-  const std::size_t l1_index = index_generator(l0_name);
+  const renamedt<ssa_exprt, L0> renamed = rename_ssa<L0>(ssa_exprt{expr}, ns);
+  const irep_idt l0_name = renamed.get_identifier();
+  const auto l1_index = narrow_cast<unsigned>(index_generator(l0_name));
 
-  const auto r_opt = level1.current_names.find(l0_name);
-
-  if(!r_opt)
-  {
-    level1.current_names.insert(l0_name, std::make_pair(ssa, l1_index));
-  }
-  else
+  if(const auto old_value = level1.insert_or_replace(renamed, l1_index))
   {
     // save old L1 name
-    if(!frame.old_level1.has_key(l0_name))
-    {
-      frame.old_level1.insert(l0_name, r_opt->get());
-    }
-
-    level1.current_names.replace(l0_name, std::make_pair(ssa, l1_index));
+    if(!frame.old_level1.has(renamed))
```

Comment:
You mean if it's present?
Instead `sharing_mapt` has a precondition what ensures we never insert twice the same key, it seems more secure to abort the program there rather than silently doing nothing. The alternative solution would be to give another name to insert.

Reply:
Ah crap forgot this was `sharing_map`. Indeed it surely ought to have the same semantics as `std::map::insert` and its current `insert` should be called something like `insert_fresh`

Reply:
It makes sense but I would leave that for another pull request since this is related more to sharing map than renamings